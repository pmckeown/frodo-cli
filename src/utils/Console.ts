/* eslint-disable no-console */
import { MultiBar, Presets } from 'cli-progress';
import { createSpinner } from 'nanospinner';
import Table from 'cli-table3';
import { frodo, state } from '@rockcarver/frodo-lib';
import Color from 'colors';

Color.enable();

let multiBarContainer = null;
let progressBar = null;
let spinner = null;

/**
 * Output a message in default color to stdout or append to `state.getOutputFile()`
 * @param {string | object} message the message
 */
function data(message: string | object, newline = true) {
  if (!message) return;
  if (state.getOutputFile()) {
    if (typeof message === 'object') {
      message = JSON.stringify(message, null, 2);
    }
    if (newline) {
      message += '\n';
    }
    frodo.utils.impex.appendTextToFile(message, state.getOutputFile());
  } else if (typeof message === 'object') {
    console.dir(message, { depth: 3 });
  } else if (newline) {
    console.log(message);
  } else {
    process.stdout.write(message);
  }
}

/**
 * Output a default color message to stderr
 * @param {Object} message the message
 */
function text(message: string | object, newline = true) {
  if (!message) return;
  if (typeof message === 'object') {
    console.dir(message, { depth: 3 });
  } else if (newline) {
    console.error(message);
  } else {
    process.stderr.write(message);
  }
}

/**
 * Output a message in bright cyan to stderr
 * @param {Object} message the message
 */
function info(message: string | object, newline = true) {
  if (!message) return;
  if (typeof message === 'object') {
    console.dir(message, { depth: 3 });
  } else if (newline) {
    console.error(message['brightCyan']);
  } else {
    process.stderr.write(message['brightCyan']);
  }
}

/**
 * Output a message in yellow to stderr
 * @param {Object} message the message
 */
function warn(message: string | object, newline = true) {
  if (!message) return;
  if (typeof message === 'object') {
    console.dir(message, { depth: 3 });
  } else if (newline) {
    console.error(message['yellow']);
  } else {
    process.stderr.write(message['yellow']);
  }
}

/**
 * Output a message in bright red to stderr
 * @param {Object} message the message
 */
function error(message: string | object, newline = true) {
  if (!message) return;
  if (typeof message === 'object') {
    console.dir(message, { depth: 3 });
  } else if (newline) {
    console.error(message['brightRed']);
  } else {
    process.stderr.write(message['brightRed']);
  }
}

/**
 * Output a debug message
 * @param {string | object} message the message
 */
function debug(message: string | object, newline = true) {
  if (!message) return;
  if (typeof message === 'object') {
    console.dir(message, { depth: 6 });
  } else if (newline) {
    console.error(message['brightMagenta']);
  } else {
    process.stderr.write(message['brightMagenta']);
  }
}

/**
 * Output a curlirize message
 * @param {string} message the message
 */
function curlirize(message: string) {
  if (!message) return;
  console.error(message['brightBlue']);
}

/**
 * Output a message in default color to stderr
 * @param {Object} message the message
 */
export function verboseMessage(message) {
  if (!message) return;
  if (state.getVerbose()) {
    text(message);
  }
}

/**
 * Output a debug message
 * @param {Object} message the message
 */
export function debugMessage(message) {
  if (!message) return;
  if (state.getDebug()) {
    debug(message);
  }
}

/**
 * Output a curlirize message
 * @param {Object} message the message
 */
export function curlirizeMessage(message) {
  if (!message) return;
  if (state.getCurlirize()) {
    curlirize(message);
  }
}

/**
 * Prints a string message to console
 *
 * @param {string} message The string message to print
 * @param {string} [type=text] "text", "info", "warn", "error" or "data". All but
 * type="data" will be written to stderr.
 * @param {boolean} [newline=true] Whether to add a newline at the end of message
 *
 */
export function printMessage(message, type = 'text', newline = true) {
  switch (type) {
    case 'data':
      data(message, newline);
      break;
    case 'text':
      text(message, newline);
      break;
    case 'info':
      info(message, newline);
      break;
    case 'warn':
      warn(message, newline);
      break;
    case 'error':
      error(message, newline);
      break;
    default:
      text(message, newline);
  }
}

/**
 * Creates a progress bar on stderr
 *
 * Example:
 * [========================================] 100% | 49/49 | Analyzing journey - transactional_auth
 *
 * @param {Number} total The total number of entries to track progress for
 * @param {String} message optional progress bar message
 * @param {Object} options progress bar configuration options
 *
 */
export function createProgressBar(
  total,
  message = null,
  options = {
    format: '[{bar}] {percentage}% | {value}/{total} | {data}',
    noTTYOutput: true,
  }
) {
  let opt = options;
  if (message == null) {
    opt = {
      format: '[{bar}] {percentage}% | {value}/{total}',
      noTTYOutput: true,
    };
  }
  // progressBar = new SingleBar(opt, Presets.legacy); // create only when needed
  // progressBar.start(total, 0, {
  //   data: message,
  // });
  multiBarContainer = new MultiBar(opt, Presets.legacy);
  progressBar = multiBarContainer.create(total, 0, {
    data: message,
  });
}

/**
 * Updates the progress bar by 1
 * @param {string} message optional message to update the progress bar
 *
 */
export function updateProgressBar(message = null) {
  if (message)
    progressBar.increment({
      data: message,
    });
  else progressBar.increment();
}

/**
 * Stop and hide the progress bar
 * @param {*} message optional message to update the progress bar
 */
export function stopProgressBar(message = null) {
  if (message)
    progressBar.update({
      data: message,
    });
  // progressBar.stop();
  multiBarContainer.stop();
  multiBarContainer = null;
}

/**
 * Create the spinner
 * @param {String} message optional spinner message
 */
export function showSpinner(message) {
  spinner = createSpinner(message).start();
}

/**
 * Stop the spinner
 * @param {String} message optional message to update the spinner
 */
export function stopSpinner(message = null) {
  if (spinner) {
    let options = {};
    if (message) options = { text: message };
    spinner.stop(options);
  }
}

/**
 * Succeed the spinner. Stop and render success checkmark with optional message.
 * @param {String} message optional message to update the spinner
 */
export function succeedSpinner(message = null) {
  if (spinner) {
    let options = {};
    if (message) options = { text: message };
    spinner.success(options);
  }
}

/**
 * Warn the spinner
 * @param {String} message optional message to update the spinner
 */
export function warnSpinner(message = null) {
  if (spinner) {
    let options = {};
    if (message) options = { text: message };
    spinner.warn(options);
  }
}

/**
 * Fail the spinner
 * @param {String} message optional message to update the spinner
 */
export function failSpinner(message = null) {
  if (spinner) {
    let options = {};
    if (message) options = { text: message };
    spinner.error(options);
  }
}

/**
 * Spin the spinner
 * @param {String} message optional message to update the spinner
 */
export function spinSpinner(message = null) {
  if (spinner) {
    let options = {};
    if (message) options = { text: message };
    spinner.update(options);
    spinner.spin();
  }
}

export function createProgressIndicator(
  type = 'determinate',
  total = 0,
  message = null
) {
  if (type === 'determinate') {
    createProgressBar(total, message);
  } else {
    showSpinner(message);
  }
}

export function updateProgressIndicator(message) {
  if (!progressBar) {
    spinSpinner(message);
  } else {
    updateProgressBar(message);
  }
}

export function stopProgressIndicator(message, status = 'none') {
  if (!progressBar) {
    switch (status) {
      case 'none':
        stopSpinner(message);
        break;
      case 'success':
        succeedSpinner(message);
        break;
      case 'warn':
        warnSpinner(message);
        break;
      case 'fail':
        failSpinner(message);
        break;
      default:
        stopSpinner(message);
        break;
    }
  } else {
    stopProgressBar(message);
  }
}

/**
 * Create an empty table
 * @param {string[]} head header row as an array of strings
 * @returns {any} an empty table
 */
export function createTable(head) {
  const table = new Table({
    head,
    chars: {
      top: '',
      'top-mid': '',
      'top-left': '',
      'top-right': '',
      bottom: '',
      'bottom-mid': '',
      'bottom-left': '',
      'bottom-right': '',
      left: '',
      'left-mid': '',
      mid: '',
      'mid-mid': '',
      right: '',
      'right-mid': '',
    },
    style: { 'padding-left': 0, 'padding-right': 0, head: ['brightCyan'] },
  });
  return table;
}

/**
 * Create a new key/value table
 * @returns {any} an empty key/value table
 */
export function createKeyValueTable() {
  const table = new Table({
    chars: {
      top: '',
      'top-mid': '',
      'top-left': '',
      'top-right': '',
      bottom: '',
      'bottom-mid': '',
      'bottom-left': '',
      'bottom-right': '',
      left: '',
      'left-mid': '',
      mid: '',
      'mid-mid': '',
      right: '',
      'right-mid': '',
    },
    style: { 'padding-left': 0, 'padding-right': 0 },
    wordWrap: true,
  });
  return table;
}

/**
 * Helper function to determine the total depth of an object
 * @param {Object} object input object
 * @returns {Number} total depth of the input object
 */
function getObjectDepth(object) {
  return Object(object) === object
    ? 1 + Math.max(-1, ...Object.values(object).map(getObjectDepth))
    : 0;
}

/**
 * Helper function to determine if an object has values
 * @param {Object} object input object
 * @returns {boolean} true of the object or any of its sub-objects contain values, false otherwise
 */
function hasValues(object) {
  let has = false;
  const keys = Object.keys(object);
  for (const key of keys) {
    if (Object(object[key]) !== object[key]) {
      return true;
    }
    has = has || hasValues(object[key]);
  }
  return has;
}

/**
 * Helper function (recursive) to add rows to an object table
 * @param {object} object object to render
 * @param {number} depth total depth of initial object
 * @param {number} level current level
 * @param {any} table the object table to add the rows to
 * @param {Object} keyMap optional JSON map to map raw config names to human readable names {'raw': 'readable'}
 * @returns the updated object table
 */
function addRows(object, depth, level, table, keyMap) {
  const space = '  ';
  const keys = Object.keys(object);
  for (const key of keys) {
    if (Object(object[key]) !== object[key]) {
      if (level === 1) {
        table.push([
          keyMap[key] ? keyMap[key].brightCyan : key['brightCyan'],
          object[key],
        ]);
      } else {
        table.push([
          {
            hAlign: 'right',
            content: keyMap[key] ? keyMap[key].gray : key.gray,
          },
          object[key],
        ]);
      }
    }
  }
  for (const key of keys) {
    if (Object(object[key]) === object[key]) {
      // only print header if there are any values below
      if (hasValues(object[key])) {
        let indention = new Array(level).fill(space).join('');
        if (level < 3) indention = `\n${indention}`;
        table.push([
          indention.concat(
            keyMap[key] ? keyMap[key].brightCyan : key['brightCyan']
          ),
          '',
        ]);
      }
      // eslint-disable-next-line no-param-reassign
      table = addRows(object[key], depth, level + 1, table, keyMap);
    }
  }
  return table;
}

/**
 * Create and populate an object table from any JSON object. Use for describe commands.
 * @param {Object} object JSON object to create
 * @param {Object} keyMap optional JSON map to map raw config names to human readable names {'raw': 'readable'}
 * @returns {any} a table that can be printed to the console
 */
export function createObjectTable(object, keyMap = {}) {
  // eslint-disable-next-line no-param-reassign
  const depth = getObjectDepth(object);
  // eslint-disable-next-line no-param-reassign
  const level = 0;
  // eslint-disable-next-line no-param-reassign
  const table = new Table({
    chars: {
      top: '',
      'top-mid': '',
      'top-left': '',
      'top-right': '',
      bottom: '',
      'bottom-mid': '',
      'bottom-left': '',
      'bottom-right': '',
      left: '',
      'left-mid': '',
      mid: '',
      'mid-mid': '',
      right: '',
      'right-mid': '',
    },
    style: { 'padding-left': 0, 'padding-right': 0, head: ['brightCyan'] },
  });
  addRows(object, depth, level + 1, table, keyMap);
  return table;
}
