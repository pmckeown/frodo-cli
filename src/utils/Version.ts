import fs from 'fs';
import path from 'path';
import os from 'os';
import { fileURLToPath } from 'url';
import { compareVersions } from 'compare-versions';
import { frodo } from '@rockcarver/frodo-lib';
import Color from 'colors';

const VERSION_CACHE_FILE = `${os.homedir()}/.frodo/Versions.json`;
const VERSION_CHECK_INTERVAL = 3600;

const GITHUB_REPOS_URL = `https://api.github.com`;
const GITHUB_RELEASES_PATH_CLI = `/repos/rockcarver/frodo-cli/releases`;

const NPM_BASE_URL = `https://registry.npmjs.org`;
const NPM_PACKAGE_PATH_CLI = `/@rockcarver/frodo-cli`;

const ENDPOINTS = [
  {
    base: GITHUB_REPOS_URL,
    path: GITHUB_RELEASES_PATH_CLI,
  },
  {
    base: NPM_BASE_URL,
    path: NPM_PACKAGE_PATH_CLI,
  },
];

let versionObject = {
  last_checked: 0,
  github: '',
  npm: '',
};

const __dirname = path.dirname(fileURLToPath(import.meta.url));

Color.enable();

const pkg = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, '../../package.json'), 'utf8')
);

function getCliVersion() {
  return `${pkg.version}`;
}

function getLibVersion() {
  return frodo.utils.version.getVersion();
}

function extractGithubReleaseInfo(data) {
  const release = {
    type: 'github',
    name: '',
    published_at: '',
  };
  // find first stable release
  const r = data.find((rel) => rel.prerelease == false);
  release.name = r.name;
  release.published_at = r.published_at;
  return release;
}

function extractNpmReleaseInfo(data) {
  const release = {
    type: 'npm',
    name: '',
    published_at: '',
  };
  // stable release
  release.name = data[`dist-tags`].latest;
  release.published_at = data.time[data[`dist-tags`].latest];
  return release;
}

async function getRemoteVersionData() {
  let useCache = true;
  if (fs.existsSync(VERSION_CACHE_FILE)) {
    const data = fs.readFileSync(VERSION_CACHE_FILE, 'utf8');
    versionObject = JSON.parse(data);
    if (
      versionObject.last_checked + VERSION_CHECK_INTERVAL <
      Math.floor(Date.now() / 1000)
    ) {
      useCache = false;
    }
  } else {
    useCache = false;
  }

  if (!useCache) {
    const allVersions = [];
    const result = await frodo.utils.version.getAllVersions(ENDPOINTS);
    result.forEach((item) => {
      if (Array.isArray(item['value'].data)) {
        allVersions.push(extractGithubReleaseInfo(item['value'].data));
      } else {
        allVersions.push(extractNpmReleaseInfo(item['value'].data));
      }
    });
    // const allVersions = await LibVersion.getAllVersions(ENDPOINTS);
    allVersions.forEach((element) => {
      // cli
      if (element.type == 'github') {
        versionObject.github = element.name;
      } else {
        versionObject.npm = element.name;
      }
    });
    versionObject.last_checked = Math.floor(Date.now() / 1000);
    fs.writeFileSync(
      VERSION_CACHE_FILE,
      JSON.stringify(versionObject, null, 2)
    );
  }
}

function getBinaryName() {
  return path.basename(process.execPath);
}

export async function getVersions(checkOnly: boolean) {
  let updateAvailable = false;
  let usingBinary = false;
  if (getBinaryName() == 'frodo') {
    usingBinary = true;
  }
  try {
    await getRemoteVersionData();
  } catch (e) {
    // Do not report error if remote version can not be obtained.
    // Just silently report installed version
    versionObject = {
      last_checked: 0,
      github: null,
      npm: null,
    };
  }

  let versionString = `You seem to be running the ${
    usingBinary ? 'binary' : 'NPM'
  } package`;

  versionString += `\nInstalled versions:`;
  versionString += `\ncli: v${getCliVersion()}\nlib: v${getLibVersion()}\nnode: ${
    process.version
  }`;
  let newVersionString = '';
  if (
    (usingBinary &&
      versionObject.github != null &&
      compareVersions(getCliVersion(), versionObject.github) == -1) ||
    (!usingBinary &&
      versionObject.npm != null &&
      compareVersions(getCliVersion(), versionObject.npm) == -1)
  ) {
    updateAvailable = true;
    newVersionString += `\n\nNewer version(s) available`;
    if (usingBinary) {
      newVersionString +=
        compareVersions(getCliVersion(), versionObject.github) == -1
          ? `\ncli (github): v${versionObject.github}`.green
          : ``;
    } else {
      newVersionString +=
        compareVersions(getCliVersion(), versionObject.npm) == -1
          ? `\ncli (npm): v${versionObject.npm}`.green
          : ``;
    }
  }
  if (checkOnly) {
    if (updateAvailable) {
      return `A new version of frodo is available.\nPlease run 'frodo -v' for more details.\n`
        .green;
    } else {
      return ``;
    }
  } else {
    return versionString + newVersionString;
  }
}
