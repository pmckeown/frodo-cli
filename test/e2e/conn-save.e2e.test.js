import cp from 'child_process';
import { promisify } from 'util';
import { removeAnsiEscapeCodes } from './utils/TestUtils';
import { connection as c } from './utils/TestConfig';
import { writeFileSync, rmSync } from 'fs';

const exec = promisify(cp.exec);

process.env['FRODO_MOCK'] = '1';
const env = {
  env: process.env,
};

const jwkFile = 'test/fs_tmp/conn-save-jwk.json';

beforeAll(() => {
  writeFileSync(jwkFile, c.saJwk);
});

afterAll(() => {
  rmSync(jwkFile);
});

describe('frodo conn save', () => {
  test(`"frodo conn save ${c.host} ${c.user} ${c.pass}": save new connection profile with existing service account and without admin account.`, async () => {
    const CMD = `frodo conn save ${c.host} ${c.user} ${c.pass}`;
    const { stderr } = await exec(CMD, env);
    expect(removeAnsiEscapeCodes(stderr)).toMatchSnapshot();
  });

  test(`"frodo conn save --sa-id ${c.saId} --sa-jwk-file ${jwkFile} ${c.host}": save new connection profile with existing service account and without admin account.`, async () => {
    const CMD = `frodo conn save --sa-id ${c.saId} --sa-jwk-file ${jwkFile} ${c.host}`;
    const { stderr } = await exec(CMD, env);
    expect(removeAnsiEscapeCodes(stderr)).toMatchSnapshot();
  });
});
