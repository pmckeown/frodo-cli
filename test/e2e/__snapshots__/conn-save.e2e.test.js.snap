// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frodo conn save "frodo conn save --sa-id b672336b-41ef-428d-ae4a-e0c082875377 --sa-jwk-file test/fs_tmp/conn-save-jwk.json https://openam-frodo-dev.forgeblocks.com/am": save new connection profile with existing service account and without admin account. 1`] = `
"Connected to https://openam-frodo-dev.forgeblocks.com/am [alpha] as service account frodo-test [b672336b-41ef-428d-ae4a-e0c082875377]
- Validating service account b672336b-41ef-428d-ae4a-e0c082875377...
âœ” Successfully validated service account b672336b-41ef-428d-ae4a-e0c082875377.
Validated and added service account with id b672336b-41ef-428d-ae4a-e0c082875377 to profile.
Saved connection profile https://openam-frodo-dev.forgeblocks.com/am
"
`;

exports[`frodo conn save "frodo conn save https://openam-frodo-dev.forgeblocks.com/am volker.scheuber@forgerock.com Sup3rS3cr3t!": save new connection profile with existing service account and without admin account. 1`] = `
"Connected to https://openam-frodo-dev.forgeblocks.com/am [alpha] as user volker.scheuber@forgerock.com
Created and added service account Frodo-SA-1687293271764 with id 17051145-f82a-47b0-8e73-91f27e2cb622 to profile.
Created log API key 2e3e7162c6861fdc50a10b6353ada65a and secret.
Saved connection profile https://openam-frodo-dev.forgeblocks.com/am
"
`;
